### PR TITLE
REFACTOR: Move `pushMetrics` calls to APIs

### PR DESCRIPTION
**Pull request by bennettbuchanan**
Refactor the former `pushMetrics` call to a new function `pushMetric` which calls the updated `UtapiClient` method of the same name (as introduced by this PR: https://github.com/scality/utapi/pull/72).

That's cool, but why also move the calls to the APIs? 

This PR is also intended to prepare the way for the addition of other Utapi metric granularities (e.g., account-level, user-level, and service-level metrics). By moving the calls to their respective APIs, we can easily add more metric properties to the object passed in `pushMetric` (this especially applies to account-level metrics because we need canonical IDs).

Depends on Utapi branch: https://github.com/scality/utapi/tree/refactor/utapiClient.

### CODE REVIEW

File: lib/api/apiUtils/bucket/bucketCreation.js [hunk 1]
```diff
@@ -79,7 +80,15 @@ function removeTransientOrDeletedLabel(bucket, log, callback) {
     const bucketName = bucket.getName();
     bucket.removeTransientFlag();
     bucket.removeDeletedFlag();
-    return metadata.updateBucket(bucketName, bucket, log, callback);
+    return metadata.updateBucket(bucketName, bucket, log, err => {
+        if (err) {
+            callback(err);
```
Comment (rahulreddy) [inline lib/api/apiUtils/bucket/bucketCreation.js]: Can you return the callbacks here?

**Comment by rahulreddy** [comment]
SCALITY_UTAPI_BRANCH=refactor/utapiClient
@ironman-machine try

**Comment by bennettbuchanan** [comment]
SCALITY_UTAPI_BRANCH=refactor/utapiClient
@ironman-machine try

**Comment by rahulreddy** [comment]
SCALITY_UTAPI_BRANCH=refactor/utapiClient
@ironman-machine try

**Comment by bennettbuchanan** [comment]
SCALITY_UTAPI_BRANCH=refactor/utapiClient
@ironman-machine try

**Comment by bennettbuchanan** [comment]
SCALITY_UTAPI_BRANCH=refactor/utapiClient
@ironman-machine try

**Comment by bennettbuchanan** [comment]
SCALITY_UTAPI_BRANCH=refactor/utapiClient
@ironman-machine try

**Comment by bennettbuchanan** [comment]
SCALITY_UTAPI_BRANCH=refactor/utapiClient
@ironman-machine try

File: lib/api/bucketGetVersioning.js [hunk 2]
```diff
@@ -63,6 +63,10 @@ export default function bucketGetVersioning(authInfo, request, log, callback) {
         }
         const versioningConfiguration = bucket.getVersioningConfiguration();
         const xml = convertToXml(versioningConfiguration);
+        // TODO push metric for gucketGetVersioning
```
Comment (electrachong) [inline lib/api/bucketGetVersioning.js]: minor typo: bucketGetVersioning

Comment (bennettbuchanan) [inline lib/api/bucketGetVersioning.js]: Thanks! Updated.

**Comment by bennettbuchanan** [comment]
end-to-end passed here http://ci.ironmann.io/gh/scality/Integration/7760

**Comment by bennettbuchanan** [comment]
SCALITY_UTAPI_BRANCH=refactor/utapiClient
@ironman-machine try

**Comment by bennettbuchanan** [comment]
SCALITY_UTAPI_BRANCH=refactor/utapiClient
@ironman-machine try

**Comment by bennettbuchanan** [comment]
Above after rebase onto master

**Comment by bennettbuchanan** [comment]
SCALITY_UTAPI_BRANCH=refactor/utapiClient
@ironman-machine try

File: lib/api/objectGet.js [hunk 3]
```diff
@@ -42,6 +43,14 @@ function objectGet(authInfo, request, log, callback) {
         const responseMetaHeaders = collectResponseHeaders(objMD);
         // 0 bytes file
         if (objMD.location === null) {
+            let contentLength = 0;
+            if (responseMetaHeaders && responseMetaHeaders['Content-Length']) {
+                contentLength = responseMetaHeaders['Content-Length'];
+            }
```
Comment (LaurenSpiegel) [inline lib/api/objectGet.js]: is this needed?

File: lib/api/objectGet.js [hunk 4]
```diff
@@ -85,6 +94,10 @@ function objectGet(authInfo, request, log, callback) {
                     objMD['x-amz-server-side-encryption'];
             }
         }
+        pushMetric('getObject', log, {
+            bucket: bucketName,
+            newByteLength: parseInt(responseMetaHeaders['Content-Length'], 10),
```
Comment (LaurenSpiegel) [inline lib/api/objectGet.js]: shouldn't need parseInt

File: lib/api/objectPutCopyPart.js [hunk 5]
```diff
@@ -320,6 +320,10 @@ function objectPutCopyPart(authInfo, request, sourceBucket,
                     serverSideEncryption.masterKeyId;
             }
         }
+        // TODO push metric for bucketGetVersioning
+        // pushMetric('getBucketVersioning', log, {
+        //      bucket: destBucketName,
+        // });
```
Comment (LaurenSpiegel) [inline lib/api/objectPutCopyPart.js]: wrong place...

File: lib/api/objectPutPart.js [hunk 6]
```diff
@@ -218,6 +219,10 @@ export default function objectPutPart(authInfo, request, streamingV4Params, log,
             });
             return cb(err);
         }
+        pushMetric('uploadPart', log, {
+            bucket: bucketName,
+            newByteLength: size,
```
Comment (LaurenSpiegel) [inline lib/api/objectPutPart.js]: don't you need old size here if overwriting?

Comment (bennettbuchanan) [inline lib/api/objectPutPart.js]: Metrics for overwrites are currently not handled in this API, but is being addressed by PR https://github.com/scality/S3/pull/472. Accordingly, the Utapi branch this PR is dependent on is not enforcing that property for this metric. If https://github.com/scality/S3/pull/472 is merged first before this one, then I will update this.

File: lib/api/objectPutCopyPart.js (hunk 5 above)
Comment (bennettbuchanan) [inline lib/api/objectPutCopyPart.js]: Updated.

File: lib/api/objectGet.js (hunk 4 above)
Comment (bennettbuchanan) [inline lib/api/objectGet.js]: Passing `responseMetaHeaders['Content-Length']` directly.

File: lib/api/objectGet.js (hunk 3 above)
Comment (bennettbuchanan) [inline lib/api/objectGet.js]: Passing `0` to `newByteLength` directly.

File: lib/api/objectGet.js (hunk 4 above)
Comment (bennettbuchanan) [inline lib/api/objectGet.js]: Utapi will enforce integer value using assertions.

**Comment by bennettbuchanan** [comment]
SCALITY_UTAPI_BRANCH=refactor/utapiClient
@ironman-machine try

**Comment by bennettbuchanan** [comment]
end-to-end passed

**Comment by bennettbuchanan** [comment]
SCALITY_UTAPI_BRANCH=refactor/utapiClient
@ironman-machine try

**Comment by bennettbuchanan** [comment]
Squashed and dropped commits.

**Comment by bennettbuchanan** [comment]
@ironman-machine try

**Pull request merged**